src/team_lib/mhba/xccl_mhba_socket_comm.c

```diff
@@ -0,0 +1,277 @@
+    struct iovec iov[1];
+    ssize_t nbytes;
+    int *p;
+    char buf[CMSG_SPACE(sizeof(int))];
+    uint32_t handles[1] = {};
+
+    iov[0].iov_base = handles;
+    iov[0].iov_len = sizeof(handles);
+
+    xccl_mhba_debug("do_recvmsg: got len of %ld\n", iov[0].iov_len);
+    memset(buf, 0x0d, sizeof(buf));
+    cmsghdr = (struct cmsghdr *)buf;
+    cmsghdr->cmsg_len = CMSG_LEN(sizeof(int));
+    cmsghdr->cmsg_level = SOL_SOCKET;
+    cmsghdr->cmsg_type = SCM_RIGHTS;
+    msg.msg_name = NULL;
+    msg.msg_namelen = 0;
+    msg.msg_iov = iov;
+    msg.msg_iovlen = sizeof(iov) / sizeof(iov[0]);
+    msg.msg_control = cmsghdr;
+    msg.msg_controllen = CMSG_LEN(sizeof(int));
+    msg.msg_flags = 0;
+
+    nbytes = recvmsg(sock, &msg, 0);
+    if (nbytes == -1){
+        xccl_mhba_error("Failed to recv msg");
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    p = (int *)CMSG_DATA(cmsghdr);
+    xccl_mhba_debug("recvmsg: %d, nbytes=%d, handle1=%d\n", *p, (int)nbytes, handles[0]);
+
+    *shared_cmd_fd = *p;
+    *shared_pd_handle = handles[0];
+
+    return XCCL_OK;
+}
+
+static xccl_status_t client_recv_data(int* shared_cmd_fd,uint32_t *shared_pd_handle)
+{
+    int sock;
+    struct sockaddr_storage sockaddr = {};
+    struct sockaddr_un *addr;
+
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1)
+    {
+        xccl_mhba_error("Failed to create client socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    addr = (struct sockaddr_un *)&sockaddr;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, SOCKPATH, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) -1 ] = '\0';
+
+    if (connect(sock, (struct sockaddr *)addr, SUN_LEN(addr)) == -1) {
+        xccl_mhba_error("Failed to connect client socket errno %d", errno);
+        goto fail;
+    }
+    if (do_recvmsg(sock, shared_cmd_fd, shared_pd_handle) != XCCL_OK) {
+        xccl_mhba_error("Failed to recv msg");
+        goto fail;
+    }
+
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close client socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    return XCCL_OK;
+
+fail:
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close client socket errno %d", errno);
+    }
+    return XCCL_ERR_NO_MESSAGE;
+}
+
+static xccl_status_t server_send_data(int command_fd, uint32_t pd_handle, int num_of_connections)
+{
+    int sock, i, num_of_curr_connections = 0;
+
+    struct sockaddr_storage storage= {};
+    struct sockaddr_un *addr;
+
+    connection_t connection[num_of_connections];
+    pthread_t thread[num_of_connections];
+
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1)
+    {
+        xccl_mhba_error("Failed to create server socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    addr = (struct sockaddr_un *)&storage;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, SOCKPATH, sizeof(addr->sun_path));
```

Comment:
using hardcoded SOCKPATH is bad (e.g., if the COMM_WORLD is split into 2 teams with different ranks, they both are trying to setup shared pd, they will use same name->crash). Need to use mktmp family of routines (https://www.gnu.org/software/libc/manual/html_node/Temporary-Files.html) to generate the uniq file in /tmp and then communicate the name from asr to other ranks on node.  We already have that oob broadcast in the team_create_post, Just need to generate the /tmp/xxx file before, and broadcast it in the same call.